File: src/test/java/at/aamio/Interop.java
package at.aamio;

import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * The Java side of tests/interop.py: reads one JSON object on stdin, opens
 * what Python sealed, verifies what Python signed, seals and signs for
 * Python, and prints one JSON object. Nothing touches the network.
 */
public final class Interop {
    private Interop() {
    }

    public static void main(String[] args) throws IOException {
        Map<String, Object> in = Json.object(Codec.utf8(System.in.readAllBytes()));
        Keys keys = Keys.fromSeedHex((String) in.get("java_seed"));
        String py = (String) in.get("py_public");
        String input = (String) in.get("signing_input");
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("java_public", keys.publicKey());
        out.put("opened", Codec.utf8(keys.open(py, (String) in.get("envelope_from_py"))));
        out.put("py_signature_verifies", Keys.verify(py, (String) in.get("py_signature"), input));
        out.put("envelope_from_java", keys.seal(py, (String) in.get("plaintext_for_py")));
        out.put("java_signature", keys.sign(input));
        System.out.write(Codec.utf8(Json.write(out) + "\n"));
        System.out.flush();
    }
}

File: tests/interop.py
"""Java and Python as one client: what one seals the other opens, what one signs the other verifies.

Run from aamio-java after `python tools/build.py`, with the aamio-python
checkout beside it and java on PATH, or AAMIO_JDK / JAVA_HOME pointing at a
JDK, or AAMIO_JAVA at the java binary:

    python tests/interop.py

Nothing touches the network. Python is the reference because its box is
PyNaCl's, which the shared vectors were made with; Java is the port under test.
"""

import json
import os
import subprocess
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.join(HERE, "..")
sys.path.insert(0, os.path.join(ROOT, "..", "aamio-python", "src"))

from nacl.signing import SigningKey, VerifyKey  # noqa: E402
from aamio.crypto import Keys, b64url, thread_signing_input, unb64url  # noqa: E402

JDK = os.environ.get("AAMIO_JDK") or os.environ.get("JAVA_HOME")
JAVA = os.environ.get("AAMIO_JAVA") or (os.path.join(JDK, "bin", "java") if JDK else "java")
CLASSPATH = os.pathsep.join([os.path.join(ROOT, "build", "classes"), os.path.join(ROOT, "build", "test-classes")])


def main():
    py = Keys(os.urandom(32))
    java_seed = os.urandom(32)
    java_public = b64url(bytes(SigningKey(java_seed).verify_key))
    signing_input = thread_signing_input("ohcibx4t22xc6hx22fch", '{"hello":"from python"}')
    request = {
        "java_seed": java_seed.hex(),
        "py_public": py.public,
        "signing_input": signing_input,
        "py_signature": py.sign(signing_input),
        "plaintext_for_py": "fra java, åpnet i python ☕",
        "envelope_from_py": py.seal(java_public, "fra python, åpnet i java 🐍".encode("utf-8")),
    }
    run = subprocess.run([JAVA, "-cp", CLASSPATH, "at.aamio.Interop"], cwd=ROOT, input=json.dumps(request).encode("utf-8"), capture_output=True, check=True)
    out = json.loads(run.stdout.decode("utf-8"))

    checks = [
        (out["java_public"] == java_public, "Java derives the same public key from the seed as PyNaCl"),
        (out["opened"] == "fra python, åpnet i java 🐍", "Java opens what Python sealed to it"),
        (out["py_signature_verifies"] is True, "Java verifies Python's signature"),
        (py.open(java_public, out["envelope_from_java"]).decode("utf-8") == request["plaintext_for_py"], "Python opens what Java sealed to it"),
    ]
    try:
        VerifyKey(unb64url(java_public)).verify(signing_input.encode("utf-8"), unb64url(out["java_signature"]))
        checks.append((True, "Python verifies Java's signature"))
    except Exception:
        checks.append((False, "Python verifies Java's signature"))

    failed = 0
    for ok, label in checks:
        print(("  ok    " if ok else "  FAIL  ") + label)
        failed += 0 if ok else 1
    print("\n%d passed, %d failed" % (len(checks) - failed, failed))
    return 1 if failed else 0


if __name__ == "__main__":
    sys.exit(main())
